File: frontend/src/api/tenantHrConfig.ts
import client from './client';

export type EstablishmentEnforcementMode = 'Off' | 'Advisory' | 'Enforced';

/**
 * Tenant HR configuration (api/tenant-hr-config).
 *
 * IMPORTANT: the PUT replaces ALL fields (same trap class as Render env vars) —
 * always round-trip the full GET payload and change only the field you mean to.
 */
export interface TenantHrConfigDto {
  useDeptHeadApproval: boolean;
  useHrFinalApproval: boolean;
  useSupervisorBeforeManager: boolean;
  allowDottedLineApproval: boolean;
  autoCreateDeptOnImport: boolean;
  autoCreateDesignationOnImport: boolean;
  requireImportPreviewBeforeCommit: boolean;
  allowCrossDeptManager: boolean;
  allowCrossLocationManager: boolean;
  requireCostCenterForPayroll: boolean;
  requireGradeForApprovalPolicy: boolean;
  /** Off | Advisory | Enforced — default Enforced (deploy-safe: absent budget rows enforce nothing). */
  establishmentEnforcementMode: EstablishmentEnforcementMode;
}

const FIELDS: (keyof TenantHrConfigDto)[] = [
  'useDeptHeadApproval', 'useHrFinalApproval', 'useSupervisorBeforeManager', 'allowDottedLineApproval',
  'autoCreateDeptOnImport', 'autoCreateDesignationOnImport', 'requireImportPreviewBeforeCommit',
  'allowCrossDeptManager', 'allowCrossLocationManager', 'requireCostCenterForPayroll',
  'requireGradeForApprovalPolicy', 'establishmentEnforcementMode',
];

export const tenantHrConfigApi = {
  get: () =>
    client.get<TenantHrConfigDto>('/api/tenant-hr-config').then(r => r.data),

  /**
   * Sends the complete request body (PUT replaces all fields — pass a full, freshly
   * fetched config). `reason` is mandatory server-side when the establishment
   * enforcement mode changes (the change is audited with actor + before/after).
   */
  update: (config: TenantHrConfigDto, reason?: string) => {
    const body: Record<string, unknown> = {};
    for (const f of FIELDS) body[f] = config[f];
    if (reason) body.reason = reason;
    return client.put<TenantHrConfigDto>('/api/tenant-hr-config', body).then(r => r.data);
  },
};

File: backend-dotnet/Zayra.Api/Domain/Entities/TenantHrConfig.cs
namespace Zayra.Api.Domain.Entities;

public class TenantHrConfig : ITenantOwned
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid TenantId { get; set; }
    // Approval routing
    public bool UseDeptHeadApproval { get; set; } = true;
    public bool UseHrFinalApproval { get; set; } = true;
    public bool UseSupervisorBeforeManager { get; set; } = false;
    public bool AllowDottedLineApproval { get; set; } = false;
    // Import behaviour
    public bool AutoCreateDeptOnImport { get; set; } = false;
    public bool AutoCreateDesignationOnImport { get; set; } = false;
    public bool RequireImportPreviewBeforeCommit { get; set; } = true;
    // Hierarchy rules
    public bool AllowCrossDeptManager { get; set; } = true;
    public bool AllowCrossLocationManager { get; set; } = true;
    // Payroll / compliance
    public bool RequireCostCenterForPayroll { get; set; } = false;
    public bool RequireGradeForApprovalPolicy { get; set; } = false;
    // Establishment matrix enforcement: Off | Advisory | Enforced. Default Enforced is
    // deploy-safe because enforcement is per-budget-row opt-in (absent row = uncontrolled);
    // changing this value is itself permission-gated + audited (never flipped by deploys).
    public string EstablishmentEnforcementMode { get; set; } = "Enforced";
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAtUtc { get; set; } = DateTime.UtcNow;
}
